### ISSUE TITLE
ValueError: cannot reshape array of size 110592 into shape (96,3,480)

### ISSUE THREAD

**Issue by tahirahmad2030** [explosion/spaCy#5221]
## How to reproduce the behaviour

I created a pretrained vector of pubmed using `python -m spacy pretrain text.jsonl /home/tahir/15-6-2018/EPI_NER/word_vectors/data/pubmed_model  ./pretrain_model --use-vectors`
 The above command created a pretraned .bin file which is say `model0.bin`

I used the same in text categorizer script `https://github.com/explosion/spaCy/blob/master/examples/training/train_textcat.py`
as 

```def main(model=None, output_dir="/home/tahir/epi_projects/RBD/model2", n_iter=10, n_texts=100000, init_tok2vec="./model0.bin", training_data='./training.csv'):```

If I don't use the pretrained embeding the script ruls fine but with `init_tok2vec="./model0.bin"`

I get the following error:
```Training the model...
LOSS 	  P  	  R  	  F  
Traceback (most recent call last):
  File "train_spacy_categorizer.py", line 214, in <module>
    plac.call(main)
  File "/home/tahir/anaconda2/envs/py36/lib/python3.6/site-packages/plac_core.py", line 328, in call
    cmd, result = parser.consume(arglist)
  File "/home/tahir/anaconda2/envs/py36/lib/python3.6/site-packages/plac_core.py", line 207, in consume
    return cmd, self.func(*(args + varargs + extraopts), **kwargs)
  File "train_spacy_categorizer.py", line 94, in main
    nlp.update(texts, annotations, sgd=optimizer, drop=0.3, losses=losses)
  File "/home/tahir/anaconda2/envs/py36/lib/python3.6/site-packages/spacy/language.py", line 519, in update
    proc.update(docs, golds, sgd=get_grads, losses=losses, **kwargs)
  File "pipes.pyx", line 986, in spacy.pipeline.pipes.TextCategorizer.update
  File "/home/tahir/anaconda2/envs/py36/lib/python3.6/site-packages/thinc/neural/_classes/feed_forward.py", line 53, in continue_update
    gradient = callback(gradient, sgd)
  File "/home/tahir/anaconda2/envs/py36/lib/python3.6/site-packages/thinc/api.py", line 300, in finish_update
    d_X = bp_layer(layer.ops.flatten(d_seqs_out, pad=pad), sgd=sgd)
  File "/home/tahir/anaconda2/envs/py36/lib/python3.6/site-packages/thinc/neural/_classes/feed_forward.py", line 53, in continue_update
    gradient = callback(gradient, sgd)
  File "/home/tahir/anaconda2/envs/py36/lib/python3.6/site-packages/thinc/api.py", line 385, in uniqued_bwd
    d_uniques = bp_Y_uniq(dY_uniq, sgd=sgd)
  File "/home/tahir/anaconda2/envs/py36/lib/python3.6/site-packages/thinc/neural/_classes/feed_forward.py", line 53, in continue_update
    gradient = callback(gradient, sgd)
  File "ops.pyx", line 128, in thinc.neural.ops.Ops.dropout.wrap_backprop.finish_update
  File "/home/tahir/anaconda2/envs/py36/lib/python3.6/site-packages/thinc/neural/_classes/layernorm.py", line 78, in finish_update
    return backprop_child(d_xhat, sgd)
  File "/home/tahir/anaconda2/envs/py36/lib/python3.6/site-packages/thinc/neural/_classes/maxout.py", line 87, in finish_update
    self.d_W += d_W.reshape((self.nO, self.nP, self.nI))
ValueError: cannot reshape array of size 110592 into shape (96,3,480)
```

In my code I printed the config:
```        if init_tok2vec is not None:
            with init_tok2vec.open("rb") as file_:
                textcat.model.tok2vec.from_bytes(file_.read())
            print("Loading pretrain")
            print(textcat.model.tok2vec)
        print(json.dumps(textcat.cfg, indent=2))```

Its output is :

```{
  "exclusive_classes": true,
  "architecture": "simple_cnn",
  "labels": [
    "Acceptable",
    "Unacceptable"
  ],
  "pretrained_vectors": null,
  "pretrained_dims": null
}
```
Let me know what should I do in this case

## Your Environment

* Operating System: Ubuntu16.4
* Python Version Used:3.6.9
* spaCy Version Used: 2.2.4
* Environment Information: Ubuntu commandline

**Comment by svlandeg** [comment]
Hi @tahirahmad2030, apologies for the late follow-up. Can you paste all the relevant code of `train_spacy_categorizer.py` ? Are you starting from a newly created `textcat` pipe component?

As you used --use-vectors in your pretrain command, you need to make sure that also the `train` command has access to use these same vectors as input. This can be done by calling `util.load_model(vectors, vocab=nlp.vocab)`, see e.g. [here](https://github.com/explosion/spaCy/blob/master/spacy/cli/train.py#L263). Because your config mentions `"pretrained_vectors": null`, I assume that is the problem.

Alternatively, you can also remove the ` --use-vectors` bit of your `pretrain` command.

Let me know whether that helps !

**Comment by tahirahmad2030** [comment]
```def main(model=None, output_dir="/home/tahir/epi_projects/RBD/error_analysis/doc_clf", n_iter=12, n_texts=200000,
         init_tok2vec='./model10', training_data='./trainin.csv'):
    if output_dir is not None:
        output_dir = Path(output_dir)
        if not output_dir.exists():
            output_dir.mkdir()

    if model is not None:
        nlp = spacy.load(model)  # load existing spaCy model
        print("Loaded model '%s'" % model)
    else:
        nlp = spacy.blank("en")  # create blank Language class
        print("Created blank 'en' model")

    # add the text classifier to the pipeline if it doesn't exist
    # nlp.create_pipe works for built-ins that are registered with spaCy
    if "textcat" not in nlp.pipe_names:
        textcat = nlp.create_pipe(
            "textcat",
            config={
                "exclusive_classes": True,
                "architecture": "ensemble",
            }
        )
        nlp.add_pipe(textcat, last=True)
    # otherwise, get it, so we can add labels to it
    else:
        textcat = nlp.get_pipe("textcat")

    # add label to text classifier
    textcat.add_label("POSITIVE")
    textcat.add_label("NEGATIVE")
    # load the IMDB dataset
    print("Loading data...")
    (train_texts, train_cats), (dev_texts, dev_cats) = load_data(training_data)
    train_texts = train_texts[:n_texts]
    train_cats = train_cats[:n_texts]
    print(
        "Using {} examples ({} training, {} evaluation)".format(
            n_texts, len(train_texts), len(dev_texts)
        )
    )
    train_data = list(zip(train_texts, [{"cats": cats} for cats in train_cats]))

    # get names of other pipes to disable them during training
    other_pipes = [pipe for pipe in nlp.pipe_names if pipe != "textcat"]
    with nlp.disable_pipes(*other_pipes):  # only train textcat
        optimizer = nlp.begin_training()
        # print(json.dumps(textcat.cfg, indent=2))
        if init_tok2vec is not None:
            with init_tok2vec.open("rb") as file_:
                textcat.model.tok2vec.from_bytes(file_.read())
            print("Loading pretrain")
            print(textcat.model.tok2vec)
        print(json.dumps(textcat.cfg, indent=2))
        print("Training the model...")
        print("{:^5}\t{:^5}\t{:^5}\t{:^5}".format("LOSS", "P", "R", "F"))
        batch_sizes = compounding(4.0, 32.0, 1.001)
        for i in range(n_iter):
            losses = {}
            # batch up the examples using spaCy's minibatch
            random.shuffle(train_data)
            batches = minibatch(train_data, size=batch_sizes)
            for batch in batches:
                texts, annotations = zip(*batch)
                nlp.update(texts, annotations, sgd=optimizer, drop=0.2, losses=losses)
            with textcat.model.use_params(optimizer.averages):
                # evaluate on the dev data split off in load_data()
                scores = evaluate(nlp.tokenizer, textcat, dev_texts, dev_cats)
            print(
                "{0:.3f}\t{1:.3f}\t{2:.3f}\t{3:.3f}".format(  # print a simple table
                    losses["textcat"],
                    scores["textcat_p"],
                    scores["textcat_r"],
                    scores["textcat_f"],
                )
            )

    # test the trained model
    test_text = "What is analysis"
    doc = nlp(test_text)
    print(test_text, doc.cats)

    if output_dir is not None:
        nlp.to_disk('./doc_clf_final/')
        with nlp.use_params(optimizer.averages):
            nlp.to_disk(output_dir)
        print("Saved model to", output_dir)

        # test the saved model
        print("Loading from", output_dir)
        nlp2 = spacy.load(output_dir)
        doc2 = nlp2(test_text)
        print(test_text, doc2.cats)

**Comment by tahirahmad2030** [comment]
def main(model=None, output_dir="/home/tahir/epi_projects/RBD/error_analysis/doc_clf", n_iter=12, n_texts=200000,
         init_tok2vec='./model10', training_data='./trainin.csv'):
    if output_dir is not None:
        output_dir = Path(output_dir)
        if not output_dir.exists():
            output_dir.mkdir()

    if model is not None:
        nlp = spacy.load(model)  # load existing spaCy model
        print("Loaded model '%s'" % model)
    else:
        nlp = spacy.blank("en")  # create blank Language class
        print("Created blank 'en' model")

    # add the text classifier to the pipeline if it doesn't exist
    # nlp.create_pipe works for built-ins that are registered with spaCy
    if "textcat" not in nlp.pipe_names:
        textcat = nlp.create_pipe(
            "textcat",
            config={
                "exclusive_classes": True,
                "architecture": "ensemble",
            }
        )
        nlp.add_pipe(textcat, last=True)
    # otherwise, get it, so we can add labels to it
    else:
        textcat = nlp.get_pipe("textcat")

    # add label to text classifier
    textcat.add_label("POSITIVE")
    textcat.add_label("NEGATIVE")
    # load the IMDB dataset
    print("Loading data...")
    (train_texts, train_cats), (dev_texts, dev_cats) = load_data(training_data)
    train_texts = train_texts[:n_texts]
    train_cats = train_cats[:n_texts]
    print(
        "Using {} examples ({} training, {} evaluation)".format(
            n_texts, len(train_texts), len(dev_texts)
        )
    )
    train_data = list(zip(train_texts, [{"cats": cats} for cats in train_cats]))

    # get names of other pipes to disable them during training
    other_pipes = [pipe for pipe in nlp.pipe_names if pipe != "textcat"]
    with nlp.disable_pipes(*other_pipes):  # only train textcat
        optimizer = nlp.begin_training()
        # print(json.dumps(textcat.cfg, indent=2))
        if init_tok2vec is not None:
            with init_tok2vec.open("rb") as file_:
                textcat.model.tok2vec.from_bytes(file_.read())
            print("Loading pretrain")
            print(textcat.model.tok2vec)
        print(json.dumps(textcat.cfg, indent=2))
        print("Training the model...")
        print("{:^5}\t{:^5}\t{:^5}\t{:^5}".format("LOSS", "P", "R", "F"))
        batch_sizes = compounding(4.0, 32.0, 1.001)
        for i in range(n_iter):
            losses = {}
            # batch up the examples using spaCy's minibatch
            random.shuffle(train_data)
            batches = minibatch(train_data, size=batch_sizes)
            for batch in batches:
                texts, annotations = zip(*batch)
                nlp.update(texts, annotations, sgd=optimizer, drop=0.2, losses=losses)
            with textcat.model.use_params(optimizer.averages):
                # evaluate on the dev data split off in load_data()
                scores = evaluate(nlp.tokenizer, textcat, dev_texts, dev_cats)
            print(
                "{0:.3f}\t{1:.3f}\t{2:.3f}\t{3:.3f}".format(  # print a simple table
                    losses["textcat"],
                    scores["textcat_p"],
                    scores["textcat_r"],
                    scores["textcat_f"],
                )
            )

    # test the trained model
    test_text = "What is analysis"
    doc = nlp(test_text)
    print(test_text, doc.cats)

    if output_dir is not None:
        nlp.to_disk('./doc_clf_final/')
        with nlp.use_params(optimizer.averages):
            nlp.to_disk(output_dir)
        print("Saved model to", output_dir)

        # test the saved model
        print("Loading from", output_dir)
        nlp2 = spacy.load(output_dir)
        doc2 = nlp2(test_text)
        print(test_text, doc2.cats)

**Comment by tahirahmad2030** [comment]
@svlandeg Thanks for you reply, The main function was used.

**Comment by tahirahmad2030** [comment]
I has also tried `"architecture": "simple_cnn",`

**Comment by tahirahmad2030** [comment]
I 



```Traceback (most recent call last):
  File "train_spacy_categorizer_test.py", line 212, in <module>
    plac.call(main)
  File "/home/ubuntu/anaconda2/envs/spacy2.2/lib/python3.6/site-packages/plac_core.py", line 367, in call
    cmd, result = parser.consume(arglist)
  File "/home/ubuntu/anaconda2/envs/spacy2.2/lib/python3.6/site-packages/plac_core.py", line 232, in consume
    return cmd, self.func(*(args + varargs + extraopts), **kwargs)
  File "train_spacy_categorizer_test.py", line 78, in main
    textcat.model.tok2vec.from_bytes(file_.read())
  File "/home/ubuntu/anaconda2/envs/spacy2.2/lib/python3.6/site-packages/thinc/neural/_classes/model.py", line 395, in from_bytes
    dest = getattr(layer, name)
AttributeError: 'FunctionLayer' object has no attribute 'G'
```

**Comment by tahirahmad2030** [comment]
I am getting below error when I try to load pretrain vectors trained in spacy 2.2.4 without using `--use-vectors`  
The error generates from line 78 where it tries to read the pretrained model.
```    with nlp.disable_pipes(*other_pipes):  # only train textcat
 75         optimizer = nlp.begin_training()
 76         if init_tok2vec is not None:
 77             with init_tok2vec.open("rb") as file_:
 78                 textcat.model.tok2vec.from_bytes(file_.read())
```

```Traceback (most recent call last):
  File "train_spacy_categorizer_test.py", line 212, in <module>
    plac.call(main)
  File "/home/ubuntu/anaconda2/envs/spacy2.2/lib/python3.6/site-packages/plac_core.py", line 367, in call
    cmd, result = parser.consume(arglist)
  File "/home/ubuntu/anaconda2/envs/spacy2.2/lib/python3.6/site-packages/plac_core.py", line 232, in consume
    return cmd, self.func(*(args + varargs + extraopts), **kwargs)
  File "train_spacy_categorizer_test.py", line 78, in main
    textcat.model.tok2vec.from_bytes(file_.read())
  File "/home/ubuntu/anaconda2/envs/spacy2.2/lib/python3.6/site-packages/thinc/neural/_classes/model.py", line 395, in from_bytes
    dest = getattr(layer, name)
AttributeError: 'FunctionLayer' object has no attribute 'G'
```

**Comment by svlandeg** [comment]
This signals that there is still an incompatibility between the Tok2Vec layer you pretrained, and the one that is used by your `textcat` component.

When you got this last error, what was the exact command used for pretraining, and which textcat architecture did you use?

You can also check the contents of your `config.json` in the folder of the pretrained model, and check the config of your pipeline with the textcat model by writing `nlp.to_disk` and checking the `cfg` file in the `textcat` folder. There's probably some parameter that's not the same.

**Comment by tahirahmad2030** [comment]
I changed the architecture to simple_cnn and the above error was solved but when I print it, it is still showing null in pretrained_vectors.

If I print this 
```if init_tok2vec is not None:
        with init_tok2vec.open("rb") as file_:
            textcat.model.tok2vec.from_bytes(file_.read())
        print("Loading pretrain")
        print(textcat.model.tok2vec)
```
I get ` <thinc.neural._classes.feed_forward.FeedForward object at 0x7f8ef95a6cc0>`

and after that I also print the below code which show it is null
```print(json.dumps(textcat.cfg, indent=2))```
I get 
```{
  "exclusive_classes": true,
  "architecture": "simple_cnn",
  "labels": [
    "Acceptable",
    "Unacceptable"
  ],
  "pretrained_vectors": null,
  "pretrained_dims": null
}
```
Is pretrained model loded. I think I am still  missing something.

**Comment by svlandeg** [comment]
Yes, that's actually as expected. By removing `--use-vectors` from your `pretrain` command, the "pretrained_vectors` setting in the Tok2Vec layer is `null`. The naming gets a bit confusing here, with all the "vectors" and all the "pretraining", but what you're printing looks good to me.

Basically what happens is this: the `pretrain` command takes vectors as input that it will try to _replicate_ by pretraining a Tok2Vec component. If you would have used `use-vectors`, those same vectors would _also_ be used as static input in the Tok2Vec architecture. But that didn't work (in your original post) because, as I explained in my original reply, you need to make sure in that case that the `train` command gets the same vectors as input to use in the static layer.

In summary: removing `--use-vectors` from your `pretrain` command and setting the architecture of the `textcat` model to `simple_cnn` resolves all issues. Right?

**Comment by tahirahmad2030** [comment]
Yes, all issues are resolved. 

I have a doubt, Do we use pretrained vectors even if `pretrained_vectors= null` and .

And it is not trivial to make sure that the train command has access to use these same vectors as input which you said can done by calling `util.load_model(vectors, vocab=nlp.vocab)` as there is no option to do it in textcat example code.

**Comment by svlandeg** [comment]
There's a difference between using `pretrained_vectors`, a static input as the first layer in a Tok2Vec component, or using a pretrained Tok2Vec component. Here, with the `pretrain` command and the way you're using the textcat example, you're doing the latter.

Happy to hear all issues are resolved - I'll go ahead and close this. Thanks again for your patience with this!

**Comment by tahirahmad2030** [comment]
Thanks @svlandeg
